pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,318 @@
+		candidateProviderIDToCommand: map[string]*Command{},
+		kubeClient:                   kubeClient,
+		recorder:                     recorder,
+		cluster:                      cluster,
+		clock:                        clock,
+	}
+	// If testing, we don't actually need to start the queue. We can call each function individually instead of relying
+	// on the rate limiting interface itself.
+	if !testingMode {
+		go queue.Start(logging.WithLogger(ctx, logging.FromContext(ctx).Named("deprovisioning")))
+	}
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, reason string, timeAdded time.Time) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates: candidates,
+		Reason:     reason,
+		TimeAdded:  timeAdded,
+	}
+}
+
+// Start will kick off the queue. This runs asynchronously until Karpenter shuts down.
+func (q *Queue) Start(ctx context.Context) {
+Loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break Loop
+		default:
+			cmd, shutdown := q.Pop()
+			if shutdown {
+				break Loop
+			}
+			q.ProcessItem(ctx, cmd)
+		}
+	}
+	logging.FromContext(ctx).Errorf("deprovisioning queue is broken and has shutdown")
+}
+
+// Pop returns a command and its hash and if the queue has shutdown.
+func (q *Queue) Pop() (*Command, bool) {
+	// Get command from queue. This waits until queue is non-empty.
+	item, shutdown := q.RateLimitingInterface.Get()
+	if shutdown {
+		return &Command{}, true
+	}
+	cmd := item.(*Command)
+	return cmd, false
+}
+
+// ProcessItem will process a command. It will either:
+// 1. Remove all references of the command if the command completes due to timeout or success.
+// 2. Check the command and requeue if it needs to be checked again later.
+func (q *Queue) ProcessItem(ctx context.Context, cmd *Command) {
+	requeue, err := q.Reconcile(ctx, cmd)
+	if !requeue {
+		// If the command timed out, log the last error.
+		if err != nil {
+			// If the command failed, bail on the action.
+			// 1. Emit metrics for launch failures
+			// 2. Ensure cluster state no longer thinks these nodes are deleting
+			// 3. Remove it from the Queue's internal data structure
+			deprovisioningReplacementNodeLaunchFailedCounter.WithLabelValues(cmd.Reason).Inc()
+			q.cluster.UnmarkForDeletion(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.ProviderID() })...)
+			err = multierr.Append(err, q.setNodesUnschedulable(ctx, false, lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.Node.Name })...))
+			logging.FromContext(ctx).Debugf("handling deprovisioning command, %w, for nodes %s", err,
+				strings.Join(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+					return s.Name()
+				}), ","))
```

Comment:
Should we use context keys instead of putting values in the string?

Reply:
Yes good callout.